File: backend/services/stigmer-server/pkg/runnerfailure/runnerfailure.go
// Package runnerfailure classifies runner activity failures at the polyglot
// Temporal boundary.
//
// The TypeScript runner and this Go control plane meet only through Temporal
// failure payloads, so "the runner's worker was shut down mid-activity" has no
// typed representation here — it arrives as one of two message shapes:
//
//  1. the activity's own classification, thrown as CancelledFailure
//     ("Activity cancelled (worker shutdown, not user pause)" — the runner's
//     execute-cursor/execute-deep-agent shutdown branches), which crosses as
//     a canceled failure; and
//  2. the Temporal TS worker's drain failing an activity that could not finish
//     inside the shutdown grace window ("Worker is shutting down and this
//     activity did not complete in time"), which crosses as a plain activity
//     failure — the shape observed live in the 2026-08-08 incident.
//
// Both mean the same thing: infrastructure interrupted the turn; the user did
// nothing. Recognizing them in ONE place keeps two behaviors consistent across
// the agentexecution and workflowexecution workflows (issue #776):
//
//   - status.error carries the honest platform-failure copy instead of raw
//     Temporal internals (the cloud channel decision table and console users
//     both read status.error);
//   - the interruption is treated as recoverable, so the workflow's bounded
//     recovery loop re-invokes from persisted harness/checkpoint state instead
//     of dead-ending the turn (owner ruling on #776).
//
// The substring matching deliberately mirrors the Java control plane's
// extractActivityFailureMessage so all editions key on the same shapes.
package runnerfailure

import (
	"errors"
	"strings"

	"go.temporal.io/sdk/temporal"
)

// WorkerShutdownStatusError is the honest status.error copy for a turn that a
// worker shutdown interrupted. Byte-identical to the copy the runner persists
// from its own shutdown branches and to the Java control plane's mapping, so
// every downstream status.error consumer keys on one string.
const WorkerShutdownStatusError = "Execution interrupted: runner worker was shut down. Retry or resume."

// canceledMarkers identify a worker shutdown inside a CANCELED failure. A
// canceled failure's message is authored by the runner's own classification
// branches (never by agent/tool output), so loose markers are safe here —
// deliberately the same two the Java control plane matches.
var canceledMarkers = []string{"worker shutdown", "shutting down"}

// drainMarker identifies the Temporal TS worker's drain text ("Worker is
// shutting down and this activity did not complete in time") in NON-canceled
// failure types. Those messages can carry agent/tool output, so the match
// anchors on the distinctive leading phrase rather than the loose markers.
const drainMarker = "worker is shutting down"

// IsWorkerShutdown reports whether err (anywhere in its chain) carries a
// runner worker-shutdown shape. Canceled failures are matched loosely (their
// messages are runner-authored); everything else must carry the Temporal
// drain phrase, because the drain shape has crossed the boundary as different
// failure types across SDK versions.
func IsWorkerShutdown(err error) bool {
	for e := err; e != nil; e = errors.Unwrap(e) {
		switch f := e.(type) {
		case *temporal.CanceledError:
			lower := strings.ToLower(f.Error())
			for _, marker := range canceledMarkers {
				if strings.Contains(lower, marker) {
					return true
				}
			}
		case *temporal.ApplicationError:
			if strings.Contains(strings.ToLower(f.Message()), drainMarker) {
				return true
			}
		default:
			if strings.Contains(strings.ToLower(e.Error()), drainMarker) {
				return true
			}
		}
	}
	return false
}

File: backend/services/stigmer-server/src/temporal/runner-failure.ts
/**
 * Runner-failure classification at the polyglot Temporal boundary — ports
 * pkg/runnerfailure (issue #776).
 *
 * "The runner's worker was shut down mid-activity" has no typed
 * representation across the Temporal boundary; it arrives as one of two
 * message shapes:
 *
 *   1. the activity's OWN classification, thrown as a cancelled failure
 *      ("Activity cancelled (worker shutdown, not user pause)" — the
 *      runner's execute-cursor/execute-deep-agent shutdown branches); and
 *   2. the Temporal TS worker's drain failing an activity that could not
 *      finish inside the shutdown grace window ("Worker is shutting down
 *      and this activity did not complete in time") — the shape observed
 *      live in the 2026-08-08 incident.
 *
 * Both mean the same thing: infrastructure interrupted the turn; the user
 * did nothing. Recognizing them in ONE place keeps two behaviors
 * consistent across the agentexecution (#18) and workflowexecution (#21)
 * workflows: status.error carries the honest platform-failure copy
 * instead of raw Temporal internals, and the interruption is treated as
 * recoverable by the bounded recovery loop (owner ruling on #776).
 *
 * Cancelled failures are matched loosely (their messages are
 * runner-authored, never agent/tool output); everything else must carry
 * the distinctive drain phrase. Deliberately the same shapes the Go and
 * Java control planes match.
 *
 * Bundle-safe: only @temporalio/common failure classes (workflow-sandbox
 * legal), imported by workflow code.
 */
import { CancelledFailure, ApplicationFailure } from "@temporalio/common";

/**
 * The honest status.error copy for a worker-shutdown-interrupted turn.
 * Byte-identical to the copy the runner persists from its own shutdown
 * branches and to the Go/Java mapping, so every downstream status.error
 * consumer keys on one string (Go runnerfailure.WorkerShutdownStatusError).
 */
export const WORKER_SHUTDOWN_STATUS_ERROR =
  "Execution interrupted: runner worker was shut down. Retry or resume.";

const CANCELED_MARKERS = ["worker shutdown", "shutting down"];

const DRAIN_MARKER = "worker is shutting down";

/**
 * Whether err (anywhere in its cause chain) carries a runner
 * worker-shutdown shape (Go runnerfailure.IsWorkerShutdown).
 */
export function isWorkerShutdown(err: unknown): boolean {
  for (
    let e: unknown = err;
    e instanceof Error;
    e = (e as { cause?: unknown }).cause
  ) {
    const lower = e.message.toLowerCase();
    if (e instanceof CancelledFailure) {
      if (CANCELED_MARKERS.some((marker) => lower.includes(marker))) {
        return true;
      }
    } else if (e instanceof ApplicationFailure) {
      if (lower.includes(DRAIN_MARKER)) {
        return true;
      }
    } else if (lower.includes(DRAIN_MARKER)) {
      return true;
    }
  }
  return false;
}
